objc_msgSend.aarch64.S

```diff
@@ -1,12 +1,78 @@
 #define ARGUMENT_SPILL_SIZE (8*10 + 8*16)
-.macro MSGSEND receiver, sel
-	.cfi_startproc
+
+/* Windows ARM64 Exception Handling
+ *
+ * Structured Exception Handling (SEH) on Windows ARM64 differs from the x64
+ * implementation. Functions consist of a single prologue and zero or more
+ * epilogues. Instead of using offsets for the .seh* directives to manipulate the
+ * stack frame, each directive corresponds to a single instruction.
+ *
+ * This presents a challenge for our objc_msgSend function, which only modifies
+ * the stack when a slow lookup is needed (see label "5").
+ *
+ * To address this, we move the directive marking the start of a function deep
+ * into the msgSend body to prevent marking every instruction as ".seh_nop."
+ *
+ *   For 64-bit Windows (_WIN64):
```

Comment:
This is a nitpick - but...

I generally very seldom see the use for checking the `_WIN64` define. If you want to check for platform/OS, then `_WIN32` is more of the canonical way. `_WIN64` was commonly used to distinguish x86_64 from i386 in Windows specific code, but when involving arm and aarch64 as well (and then it's better to explicitly check for individual architectures), it's often kinda pointless unless you explicitly want to check for the pointer size.

As this file seems to be aarch64 specific, we know the architecture already, so the OS is the only variable, so I'd just check for `_WIN32`.

Then secondly, regarding the use fo "64 bit Windows" in the text of the comment here - most of the logic for unwind info doesn't really generalize across 64 bit Windows, and where it does generalize, it's not about 32 vs 64 bit.

On i386, this kind of table based SEH unwind info isn't used at all; if building with mingw toolchains that use DWARF, then they use regular `.cfi_*` directives though.

On x86_64, `.seh_*` directives similar to on aarch64 are used - although the similarities are very few. Common ones such as `.seh_proc`, `.seh_endproc` and `.seh_stackalloc` are shared, but x86_64 doesn't have the same logic for epilogues IIRC, and all the individual directives are wildly different anyway - so I don't see much value to group them together.

On 32 bit ARM, a similar mechanism is used as on x86_64 and AArch64. The style/logic (1:1 mapping between instructions and unwind opcodes, same style with prologues/epilogues, etc) is exactly the same as for AArch64, but most of the actual opcodes differ here as well. In practice, there's little reason to care about modern Windows on 32 bit ARM (unless you care about Windows Phone, WinRT or Windows 10 IoT) though. But it's another example case against grouping unwind info style into "64 bit Windows".

I.e., TL;DR, I'd skip the wording about "64 bit Windows" and just talk about Windows in general, unless these macros actually are used in a header shared across multiple architectures.

Reply:
Thank you for the in depth answer :)